runtime/protocol/http-test/common/src/aws/smithy/kotlin/runtime/httptest/HttpTrafficParser.kt

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package aws.smithy.kotlin.runtime.httptest
+
+import aws.smithy.kotlin.runtime.http.*
+import aws.smithy.kotlin.runtime.http.content.ByteArrayContent
+import aws.smithy.kotlin.runtime.http.request.HttpRequest
+import aws.smithy.kotlin.runtime.http.request.HttpRequestBuilder
+import aws.smithy.kotlin.runtime.http.request.url
+import aws.smithy.kotlin.runtime.http.response.HttpResponse
+import aws.smithy.kotlin.runtime.util.decodeBase64Bytes
+import kotlinx.serialization.json.*
+
+internal fun parseHttpTraffic(json: String) = buildTestConnection {
+    val traffic = Json.parseToJsonElement(json).jsonArray
+    for (call in traffic.map { it.jsonObject }) {
+        val reqObj = call["request"] ?: error("expected `request` in document")
+        val respObj = checkNotNull(call["response"]) { "expected `response` key in document " }.jsonObject
+
+        val req = parseRequest(reqObj)
+        val resp = parseResponse(respObj)
+        if (req != null) {
+            expect(req, resp)
+        } else {
+            // no assertions made for null requests
+            expect(resp)
+        }
+    }
+}
+
+private enum class BodyContentType {
+    UTF_8,
+    BINARY;
+    companion object {
+        fun fromValue(value: String): BodyContentType = when (value.lowercase()) {
+            "utf8" -> UTF_8
+            "binary" -> BINARY
+            else -> error("invalid body content type: $value")
+        }
+    }
+}
+
+private fun parseRequest(req: JsonElement): HttpRequest? = when (req) {
+    // no assertions will be made
+    is JsonNull -> null
+    is JsonObject -> {
+        val method = req["method"]?.jsonPrimitive?.content ?: "GET"
```

Comment:
suggestion

FWIW there are easy ways of generating data classes from a json snippet and then you can just serialize to object form, may look cleaner

Reply:
Sorry Ian not going to happen 😉

But yeah I considered using kotlinx.serialization plugin but the headers format and http body formats aren't straightforward and this was not terribly difficult to implement without it